ethers-solc/src/lib.rs

```diff
@@ -194,15 +197,23 @@ impl<Artifacts: ArtifactOutput> Project<Artifacts> {
             if !self.allowed_lib_paths.0.is_empty() {
                 solc = solc.arg("--allow-paths").arg(self.allowed_lib_paths.to_string());
             }
+            solc_versions.insert(solc.solc.clone(), version);
             let entry = sources_by_version.entry(solc).or_insert_with(BTreeMap::new);
-            entry.insert(path, source);
+            entry.insert(path.clone(), source);
         }
 
         let mut compiled =
             ProjectCompileOutput::with_ignored_errors(self.ignored_error_codes.clone());
 
         // run the compilation step for each version
         for (solc, sources) in sources_by_version {
+            // verify that this solc version's checksum matches the checksum found remotely. If
+            // not, re-install the same version.
+            let version = solc_versions.get(&solc.solc).unwrap();
+            if let Err(_e) = solc.verify_checksum() {
```

Comment:
For a later PR, we should add a tracing log here that it found a checksum mismatch